src/github.com/matrix-org/dendrite/common/config/config.go

```diff
@@ -0,0 +1,339 @@
+		config.Media.MaxFileSizeBytes = &defaultMaxFileSizeBytes
+	}
+}
+
+func (e Error) Error() string {
+	if len(e.Problems) == 1 {
+		return e.Problems[0]
+	}
+	return fmt.Sprintf(
+		"%s (and %d other problems)", e.Problems[0], len(e.Problems)-1,
+	)
+}
+
+func (config *Dendrite) check() error {
+	var problems []string
+
+	if config.Version != Version {
+		return Error{[]string{fmt.Sprintf(
+			"unknown config version %q, expected %q", config.Version, Version,
+		)}}
+	}
+
+	checkNotEmpty := func(key string, value string) {
+		if value == "" {
+			problems = append(problems, fmt.Sprintf("missing config key %q", key))
+		}
+	}
+
+	checkNotZero := func(key string, value int64) {
+		if value == 0 {
+			problems = append(problems, fmt.Sprintf("missing config key %q", key))
+		}
+	}
+
+	checkPositive := func(key string, value int64) {
+		if value < 0 {
+			problems = append(problems, fmt.Sprintf("invalid value for config key %q: %d", key, value))
+		}
+	}
+
+	checkNotEmpty("matrix.server_name", string(config.Matrix.ServerName))
+	checkNotEmpty("matrix.private_key", string(config.Matrix.PrivateKeyPath))
+	checkNotZero("matrix.federation_certificates", int64(len(config.Matrix.FederationCertificatePaths)))
+
+	checkNotEmpty("media.base_path", string(config.Media.BasePath))
+	checkPositive("media.max_file_size_bytes", int64(*config.Media.MaxFileSizeBytes))
+	checkPositive("media.max_thumbnail_generators", int64(config.Media.MaxThumbnailGenerators))
+	for i, size := range config.Media.ThumbnailSizes {
+		checkPositive(fmt.Sprintf("media.thumbnail_sizes[%d].width", i), int64(size.Width))
+		checkPositive(fmt.Sprintf("media.thumbnail_sizes[%d].height", i), int64(size.Height))
+	}
+
+	checkNotZero("kafka.addresses", int64(len(config.Kafka.Addresses)))
+	checkNotEmpty("kafka.topics.input_room_event", string(config.Kafka.Topics.InputRoomEvent))
+	checkNotEmpty("kafka.topics.output_room_event", string(config.Kafka.Topics.OutputRoomEvent))
+	checkNotEmpty("database.media_server", string(config.Database.MediaServer))
+	checkNotEmpty("database.account", string(config.Database.Account))
+	checkNotEmpty("database.device", string(config.Database.Device))
+	checkNotEmpty("database.server_key", string(config.Database.ServerKey))
+	checkNotEmpty("database.sync_server", string(config.Database.SyncServer))
+	checkNotEmpty("database.room_server", string(config.Database.RoomServer))
+	checkNotEmpty("listen.media_api", string(config.Listen.MediaAPI))
+	checkNotEmpty("listen.client_api", string(config.Listen.ClientAPI))
+	checkNotEmpty("listen.federation_api", string(config.Listen.FederationAPI))
+	checkNotEmpty("listen.sync_api", string(config.Listen.SyncAPI))
+	checkNotEmpty("listen.room_server", string(config.Listen.RoomServer))
+
+	if problems != nil {
+		return Error{problems}
+	}
+
+	return nil
+}
+
+func absPath(dir string, path Path) string {
+	if filepath.IsAbs(string(path)) {
+		return filepath.Clean(string(path))
+	}
+	return filepath.Join(dir, string(path))
```

Comment:
Maybe comment that `filepath.Join()` calls `filepath.Clean()` on the result so both are cleaned.